File: src/main/java/dev/robothanzo/werewolf/service/RoleService.java
package dev.robothanzo.werewolf.service;

import java.util.List;
import java.util.function.Consumer;

/**
 * Service for managing the roles available in a game session
 * and handling the assignment of these roles to players.
 */
public interface RoleService {
    /**
     * Adds a specific amount of a role to the role pool for a guild.
     *
     * @param guildId  the ID of the guild
     * @param roleName the name of the role to add
     * @param amount   the amount of the role to add
     */
    void addRole(long guildId, String roleName, int amount);

    /**
     * Removes a specific amount of a role from the role pool for a guild.
     *
     * @param guildId  the ID of the guild
     * @param roleName the name of the role to remove
     * @param amount   the amount of the role to remove
     */
    void removeRole(long guildId, String roleName, int amount);

    /**
     * Retrieves the current role pool for a guild.
     *
     * @param guildId the ID of the guild
     * @return a list of role names in the pool
     */
    List<String> getRoles(long guildId);

    /**
     * Randomly assigns roles from the pool to the players in a game session.
     *
     * @param guildId          the ID of the guild
     * @param statusCallback   callback for status messages during assignment
     * @param progressCallback callback for progress percentage (0-100)
     * @throws Exception if an error occurs during assignment
     */
    void assignRoles(long guildId, Consumer<String> statusCallback, Consumer<Integer> progressCallback)
            throws Exception;
}

File: src/main/kotlin/dev/robothanzo/werewolf/service/RoleService.kt
package dev.robothanzo.werewolf.service

import dev.robothanzo.werewolf.database.documents.Session

/**
 * Service for managing the roles available in a game session
 * and handling the assignment of these roles to players.
 */
interface RoleService {
    /**
     * Adds a specific amount of a role to the role pool.
     *
     * @param session  the session to add roles to
     * @param roleName the name of the role to add
     * @param amount   the amount of the role to add
     */
    fun addRole(session: Session, roleName: String, amount: Int)

    /**
     * Removes a specific amount of a role from the role pool.
     *
     * @param session  the session to remove roles from
     * @param roleName the name of the role to remove
     * @param amount   the amount of the role to remove
     */
    fun removeRole(session: Session, roleName: String, amount: Int)

    /**
     * Randomly assigns roles from the pool to the players in a game session.
     *
     * @param session          the session
     * @param statusCallback   callback for status messages during assignment
     * @param progressCallback callback for progress percentage (0-100)
     * @throws Exception if an error occurs during assignment
     */
    @Throws(Exception::class)
    fun assignRoles(
        session: Session,
        statusCallback: (String) -> Unit,
        progressCallback: (Int) -> Unit
    )
}
